fix aot autograd for None fw inputs

bdhirsh opened the pull request:
hot fix: Confirmed this fixes an internal model that had None as one if its inputs

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89975

bdhirsh committed:
fix aot autograd for None fw inputs

[ghstack-poisoned]

aazzolini commented on test/functorch/test_aotdispatch.py:
```diff
@@ -364,6 +364,16 @@ def forward(self, primals_1):
+    def test_input_mutation_simple_with_none(self):
+        def f(a, b):
+            return a * 3
+        inp = [torch.ones(3, 3, requires_grad=True), None]
```
Could you add other varieties of non-tensor inputs as well? e.g. list<int> , int ?

aazzolini reviewed (approved):
Thanks for the fix!

ezyang commented:
I'm not sure i understand this PR. Isn't the invariant we only get Tensor inputs

aazzolini commented:
What is the status on this PR? There's a regression in some of our work that depends on this.

bdhirsh commented:
The contents of this PR (and the feedback about making the test better) is all in https://github.com/pytorch/pytorch/pull/89532. Since that PR will take some time to land, I'll land this first.

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update on "fix aot autograd for None fw inputs"

hot fix: Confirmed this fixes an internal model that had None as one if its inputs




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot rebase

bdhirsh committed:
Update on "fix aot autograd for None fw inputs"

hot fix: Confirmed this fixes an internal model that had None as one if its inputs




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -f "test failure looks unrelated https://github.com/pytorch/pytorch/actions/runs/3696505990/jobs/6260760134"
